src/cryptonote_core/service_node_list.cpp

```diff
@@ -295,70 +307,120 @@ namespace service_nodes
+    const crypto::public_key& key = state->workers[state_change.service_node_index];
 
     auto iter = m_transient_state.service_nodes_infos.find(key);
-    if (iter == m_transient_state.service_nodes_infos.end())
+    if (iter == m_transient_state.service_nodes_infos.end()) {
+      LOG_PRINT_L2("Received state change tx for non-registered service node " << key << " (perhaps a delayed tx?)");
       return false;
-
-    if (m_service_node_pubkey && *m_service_node_pubkey == key)
-    {
-      MGINFO_RED("Deregistration for service node (yours): " << key);
-    }
-    else
-    {
-      LOG_PRINT_L1("Deregistration for service node: " << key);
     }
 
-    m_transient_state.rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
+    auto &info = iter->second;
+    bool is_me = m_service_node_pubkey && *m_service_node_pubkey == key;
 
-    int hard_fork_version = m_blockchain.get_hard_fork_version(block_height);
-    if (hard_fork_version >= cryptonote::network_version_11_infinite_staking)
-    {
-      for (const auto &contributor : iter->second.contributors)
-      {
-        for (const auto &contribution : contributor.locked_contributions)
+    switch (state_change.state) {
+      case new_state::deregister:
+        if (is_me)
+          MGINFO_RED("Deregistration for service node (yours): " << key);
+        else
+          LOG_PRINT_L1("Deregistration for service node: " << key);
+
+        if (hard_fork_version >= cryptonote::network_version_11_infinite_staking)
         {
-          key_image_blacklist_entry entry = {};
-          entry.version                   = get_min_service_node_info_version_for_hf(hard_fork_version);
-          entry.key_image                 = contribution.key_image;
-          entry.unlock_height             = block_height + staking_num_lock_blocks(m_blockchain.nettype());
-          m_transient_state.key_image_blacklist.push_back(entry);
-
-          const bool adding_to_blacklist = true;
-          m_transient_state.rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_key_image_blacklist(block_height, entry, adding_to_blacklist)));
+          for (const auto &contributor : info.contributors)
+          {
+            for (const auto &contribution : contributor.locked_contributions)
+            {
+              key_image_blacklist_entry entry = {};
+              entry.version                   = get_min_service_node_info_version_for_hf(hard_fork_version);
+              entry.key_image                 = contribution.key_image;
+              entry.unlock_height             = block_height + staking_num_lock_blocks(m_blockchain.nettype());
+              m_transient_state.key_image_blacklist.push_back(entry);
+
+              const bool adding_to_blacklist = true;
+              m_transient_state.rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_key_image_blacklist(block_height, entry, adding_to_blacklist)));
```

Comment:
can you use `std::make_unique` here instead of having a raw `new` around?

Reply:
I have other code (not included in this PR) to change this (and all the others) to:
```c++
m_transient_state.rollback_events.emplace_back(new rollback_key_image_blacklist(block_height, entry, adding_to_blacklist));
```
but as to your specific question: no, because the codebase is still c++11 and make_unique was one of those sad omissions that didn't get fixed util c++14.